layers/+lang/typescript/packages.el

```diff
@@ -71,6 +72,46 @@
                          (when (configuration-layer/package-usedp 'company)
                                (company-mode-on))))))))
 
+(defun typescript/init-typescript-mode ()
+  (use-package typescript-mode
+    :defer t
+    :commands (typescript/format-buffer)
+    :config (progn
+              (defun typescript/format-buffer ()
+                "Format buffer with tsfmt."
+                (interactive)
+                (if (executable-find "tsfmt")
+                    (let*  ((tmpfile (make-temp-file "~fmt-tmp" nil ".ts"))
+                            (coding-system-for-read 'utf-8)
+                            (coding-system-for-write 'utf-8)
+                            (outputbuf (get-buffer-create "*~fmt-tmp.ts*")))
+                      (unwind-protect
+                          (progn
+                            (with-current-buffer outputbuf (erase-buffer))
+                            (write-region nil nil tmpfile)
+                            (if (zerop (apply 'call-process "tsfmt" nil outputbuf nil (list tmpfile)))
+                                (let ((p (point)))
+                                  (save-excursion
+                                    (with-current-buffer (current-buffer)
+                                      (erase-buffer)
+                                      (insert-buffer-substring outputbuf)))
+                                  (goto-char p)
+                                  (message "formatted.")
+                                  (kill-buffer outputbuf))
+                                (progn 
+                                  (message "Formatting failed!")
+                                  (display-buffer outputbuf)))
+                            (progn
+                              (delete-file tmpfile)))))
+                  (message "tsfmt not found. Run \"npm install -g typescript-formatter\"")))
+
+              (spacemacs/set-leader-keys-for-major-mode 'typescript-mode "=" 'typescript/format-buffer)
+
+              (when typescript-fmt-on-save
+                    (add-hook 'typescript-mode-hook
+                              (lambda ()
+                                (add-hook 'before-save-hook 'typescript/format-buffer t t)))))))
```

Comment:
Also, is it necessary adding to a hook a function that adds a hook? Isn't there any other way? I say it to avoid saving the same function everytime.

I'd say something like:

```elisp
(add-hook 'before-save-hook (lambda ()
                               (if (equal major-mode typescript-mode)
                                 (call-interactively 'typescritp/format-buffer))))
```

I don't know if my code is actually better at performance than the hook to hook, I just throw it to think about it.